Feature/9/create initial form

julia-rom opened the pull request:
# Description

- add PUT api request to submit registration form data
- add loading/error states for submit
- finalize API payload (almost done)
- update questions to match framework review

Fixes[ # 9](https://github.com/globalmangrovewatch/gmw-users/issues/9)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Instructions

1. `npm i`
2. `npm start`
3. fill out form
4. Open console
5. ensure no new errors are being generated and that data structure is correctly being sent through with a 200 ok response

# How Has This Been Tested?

- Tested form submissions with different inputs
- Ensured payload to axios is in the correct format (currently console logging data being sent through to axios - you should see this upon submitting)
- Ensured response is 200 OK

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

julia-rom commented:
for the answer payload, I know @ghelobytes is expecting the question_id to be in the '1.1 or 1.2a' format as an example.

The problem with this is that I will need to update the labels to be `1.1` instead of `projectTitle`. The labels are part of the response object from react hook form (more context in the` onSubmit `function). It is an unusual code pattern to have the object label be a number, and also causes issues when trying to access the object property. For example: `obj.1.1` will error. We could use obj['1.1'], however. Maybe I can just make a developer note for this, but still seems like there could be a better way of organizing this. 

Let me know if you have some ideas/thoughts @JBurkinshaw

ghelobytes commented:
> for the answer payload, I know @ghelobytes is expecting the `question_id` to be in the '1.1 or 1.2a' format as an example.
> 
> The problem with this is that I will need to update the labels to be `1.1` instead of `projectTitle`. The labels are part of the response object from react hook form (more context in the`onSubmit`function). It is an unusual code pattern to have the object label be a number, and also causes issues when trying to access the object property. For example: `obj.1.1` will error. We could use `obj['1.1']`, however. Maybe I can just make a developer note for this, but still seems like there could be a better way of organizing this.
> 
> It also makes the code harder to read and understand.
> 
> Let me know if you have some ideas/thoughts @JBurkinshaw

@julia-rom  To clarify, `question_id` will be starting with `Q` as in `Q1.1` or `Q1.2a`.

julia-rom commented:
another thing to note is that react-hook-form/yup won't allow to submit a key with a '.' in it, even as a string, so will have to think of some sort of mapping

julia-rom commented:
@JBurkinshaw I decided it'd be better to keep the form name descriptions as is, to make it easier to read, and to do the question id - name mapping as Angelo suggested. Let me know what you think

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -1,14 +1,15 @@
+import { useState } from 'react'
 import styles from './style.module.scss'
+import axios from 'axios'
 import { useForm, Controller } from 'react-hook-form'
 import { yupResolver } from '@hookform/resolvers/yup'
 import * as Yup from 'yup'
 import countries from '../../data/countries.json'
+import { projectDetailsMapping } from '../../data/questionMappingProjectDetails'
 import {
```
Can you separate _external_ imports (react, react-hook-form, yup etc) from _internal_ imports specific to this project? It makes things much easier to find and understand 🙂

julia-rom replied:
yup! will keep external at the top and internal below

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -38,68 +37,54 @@ function ProjectDetailsForm() {
+    setisSubmitting(true)
+    setIsError(false)
+    let preppedData = []
+    const url = 'https://mrtt-api-test.herokuapp.com/api/v2/sites/1/registration_answers'
```
`https://mrtt-api-test.herokuapp.com/api/v2/` should be an environment variable. That way we can ensure it's the same throughout the app and can be updated based on the environment we deploy to. See [docs](https://create-react-app.dev/docs/adding-custom-environment-variables/) for more info.

This could be done as part of a refactor when writing the next form.

julia-rom replied:
agreed! will start an env file

JBurkinshaw commented on mrtt-ui/src/data/questionMappingProjectDetails.js:
```diff
@@ -0,0 +1,7 @@
+export const projectDetailsMapping = {
+  hasProjectEndDate: '1.1a',
+  projectStartDate: '1.1b',
+  projectEndDate: '1.1c',
+  countries: '1.2',
+  siteArea: '1.3'
+}
```
I like this. As you create more forms you may want to add some nesting to the structure. Something like (or whatever is more suitable)

```js
export const projectDetailsMapping = {
  registration: {
    siteDetailsAndLocation: {
      hasProjectEndDate: '1.1a',
      projectStartDate: '1.1b',
      projectEndDate: '1.1c',
      countries: '1.2',
      siteArea: '1.3'
    },
    siteBackground: {
      //etc...
    }
  },
  intervention: {
    //etc...
  }
}
```

Or you could export them separately:
```js
export const registrationMapping = {
  siteDetailsAndLocation: {
      hasProjectEndDate: '1.1a',
      projectStartDate: '1.1b',
      projectEndDate: '1.1c',
      countries: '1.2',
      siteArea: '1.3'
    },
    siteBackground: {
      //etc...
    }
}

export const interventionMapping = {
  //etc...
}
```

julia-rom replied:
got the idea from Angelo! I was thinking of exporting them separately, but can work either or ! I'm open to either

julia-rom replied:
I think I prefer it as one object :) just updated

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -38,68 +37,54 @@ function ProjectDetailsForm() {
-  }
+    for (const [key, value] of Object.entries(data)) {
+      // map question ids with keys
+      // eslint-disable-next-line no-prototype-builtins
```
I looked at the [docs](https://eslint.org/docs/rules/no-prototype-builtins) for this.
The suggestion is
> To avoid subtle bugs like this, it's better to always call these methods from Object.prototype. For example, foo.hasOwnProperty("bar") should be replaced with Object.prototype.hasOwnProperty.call(foo, "bar").

I don't have a strong opinion on it but it might be better to take that approach to avoid bugs and please the linter. Although it definitely doesn't looks as nice.

julia-rom replied:
no problem - just updated!

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -190,8 +175,15 @@ function ProjectDetailsForm() {
+            1.3 What is the overall site area?
+          </FormLabel>
+        </div>
+        {isError && <div className={styles.invalid}>{'Submit failed, please try again'}</div>}
```
This could be simplified to
```suggestion
        {isError && <div className={styles.invalid}>Submit failed, please try again</div>}
```

julia-rom replied:
ah yes, no idea why those curly braces got added

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -190,8 +175,15 @@ function ProjectDetailsForm() {
+        </div>
+        {isError && <div className={styles.invalid}>{'Submit failed, please try again'}</div>}
+        <Button sx={{ marginTop: '1em' }} variant="contained" type="submit" disabled={isSubmitting}>
+          {isSubmitting ? `Submitting...` : `Submit`}
```
Backticks aren't required here
```suggestion
          {isSubmitting ? 'Submitting...' : 'Submit'}
```

julia-rom replied:
yup, not sure why I put those there haha

JBurkinshaw reviewed (changes requested):
Nice work. I like the `onSubmit()` function.

Looks like a lot of files from `/node_modules` have been added. I think this is because you have run `npm install` at the `gmw-users` directory level instead of within `gmw-users/mrtt-ui`. A `package-lock.json` file has been added at that level too. Can you remove these from the repository and prevent them from being committed again?

I think the node modules files got around your `.gitignore` file because it is within `gmw-users/mrtt-ui`. I also noticed there's a `.DS_Store` file in the root directory which shouldn't be there and that's probably for the same reason.

Maybe a `.gitignore` file containing `node_modules/` and whatever you think is relevant would be a good idea in the root directory (you can keep a project-specific one in `mrtt-ui` too).

JBurkinshaw commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -38,68 +37,54 @@ function ProjectDetailsForm() {
-    // POST to api route
-    console.log('data: ', data)
-  }
+    for (const [key, value] of Object.entries(data)) {
```
As the app progresses, consider if this loop and its contents can be abstracted into a helper function that can be used in all forms

julia-rom replied:
good idea! will add this folder and think about how to abstract this

julia-rom commented:
> Nice work. I like the `onSubmit()` function.
> 
> Looks like a lot of files from `/node_modules` have been added. I think this is because you have run `npm install` at the `gmw-users` directory level instead of within `gmw-users/mrtt-ui`. A `package-lock.json` file has been added at that level too. Can you remove these from the repository and prevent them from being committed again?
> 
> I think the node modules files got around your `.gitignore` file because it is within `gmw-users/mrtt-ui`. I also noticed there's a `.DS_Store` file in the root directory which shouldn't be there and that's probably for the same reason.
> 
> Maybe a `.gitignore` file containing `node_modules/` and whatever you think is relevant would be a good idea in the root directory (you can keep a project-specific one in `mrtt-ui` too).

definitely got added by mistake! added a top level git ignore and also deleted that DS_Store file

julia-rom commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -117,8 +102,8 @@ function ProjectDetailsForm() {
         <div className={styles.formGroup}>
-          <FormLabel sx={{ color: 'black', marginBottom: '1.5em' }}>Project Duration</FormLabel>
-          <FormLabel sx={{ color: 'black' }}>1.2b</FormLabel>
+          <FormLabel sx={{ color: 'black', marginBottom: '0.5em' }}>Project Duration</FormLabel>
```
yes that would be related to the draft theming PR

julia-rom commented on mrtt-ui/src/components/ProjectDetailsForm/ProjectDetailsForm.js:
```diff
@@ -190,8 +175,15 @@ function ProjectDetailsForm() {
+          </FormLabel>
+        </div>
+        {isError && <div className={styles.invalid}>{'Submit failed, please try again'}</div>}
+        <Button sx={{ marginTop: '1em' }} variant="contained" type="submit" disabled={isSubmitting}>
```
hmm I normally keep JSX as double quotes - If I do this, I get some conflicting errors. Might need to spend a bit more time on it

JBurkinshaw replied:
Ok no problem, let's leave it for now

julia-rom commented:
this is ready for a re-review. two notes:
1. I still need to spend more time thinking about a better way to create a function abstraction within onSubmit
2. Need more time to figure out JSX single quotes, as it is showing some conflicts with eslint I haven't figured out yet

@JBurkinshaw

JBurkinshaw reviewed (approved):
Approving with a few things to note.
- Maybe the root `.gitignore` should contain most of the rules that would apply to other UI apps as well, and then anything specific to mrtt-ui can live in a `mrtt-ui/.gitignore` instead?
- Can you add remove `.env` and add it to `.gitignore`? We will want to put secrets in here later so we don't want them to find their way into the repo. You could add a `.env.sample` file instead which would look like this:
  ```
  REACT_APP_API_URL=
  ```
- When you merge, can you use the "Squash Merge" option and modify the automatic suggested message to contain a summary of the changes? It can be based on previous commit messages. This will keep the commit history cleaner and easier to read by removing the work-in-progress commits. I think we should take an approach similar to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) for commit messages. I'm planning on adding more on this to the readme. For example
  ```
  [<issue no>] <feat/fix etc>: <short description>
  [body - one or more lines for longer descriptions of specifics]
  ```
  For this issue it would look something like this
  ```
    [9] feat: Create initial registration form
    Add MUI library
    Add something else
    Add another thing
    Modify such and such
  ```

julia-rom commented:
> Approving with a few things to note.
> 
> * Maybe the root `.gitignore` should contain most of the rules that would apply to other UI apps as well, and then anything specific to mrtt-ui can live in a `mrtt-ui/.gitignore` instead?
> * Can you add remove `.env` and add it to `.gitignore`? We will want to put secrets in here later so we don't want them to find their way into the repo. You could add a `.env.sample` file instead which would look like this:
>   ```
>   REACT_APP_API_URL=
>   ```
> * When you merge, can you use the "Squash Merge" option and modify the automatic suggested message to contain a summary of the changes? It can be based on previous commit messages. This will keep the commit history cleaner and easier to read by removing the work-in-progress commits. I think we should take an approach similar to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) for commit messages. I'm planning on adding more on this to the readme. For example
>   ```
>   [<issue no>] <feat/fix etc>: <short description>
>   [body - one or more lines for longer descriptions of specifics]
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   For this issue it would look something like this
>   ```
>     [9] feat: Create initial registration form
>     Add MUI library
>     Add something else
>     Add another thing
>     Modify such and such
>   ```

oops! did not mean to add that env file in - good catch. Will delete and then update gitignore.

I will use squash and merge and add more context to the message!

julia-rom commented:
did some clean up - I think we only need the .gitignore on the mrtt-ui level. I don't see the DS_Store being added anymore. @JBurkinshaw  I just haven't seen two gitignore files before and think it maybe causing some weird issues

JBurkinshaw commented:
> did some clean up - I think we only need the .gitignore on the mrtt-ui level. I don't see the DS_Store being added anymore. @JBurkinshaw I just haven't seen two gitignore files before and think it maybe causing some weird issues

@julia-rom It should be possible to have multiple gitignore files but what you've done here is fine too. We can always change later. https://stackoverflow.com/questions/9730486/can-you-have-additional-gitignore-per-directory-within-a-single-repo
